### PR TITLE
[v2]: Consolidate GPU-related CLI arguments for better clarity

### PR DESCRIPTION
**Pull request by JacksonBurns**
## Description
During an unrelated PR we started discussing how to best expose the devices for training to the CLI. This is a first pass at that. I'll go ahead and request reviews, but really just to gather input and thoughts.

## Current workflow  -> Desired workflow
All previous args related to the hardware choice (`--gpu`, `--n_gpus`, and `--no-cuda`) are now covered by `--gpu-spec`.

User can specify `-2` to use the CPU (covering `--no-cuda`), which @KnathanM indicated is a nice feature to have.
Specifying `-1` will use all GPUs, or rather all accelerators that lightning can find when using `"auto"` (which should be all of them).
Finally, in the odd case that users want to use _exactly_ one, or two, or some specific set of GPUs, they can just specify the indices of the GPUs.

## Questions
How do I run the input validation for common args? We should raise an error if the user requests a GPU but CPU is the only thing available.

### CODE REVIEW

File: chemprop/cli/common.py [hunk 1]
```diff
@@ -11,37 +13,15 @@
...
-    data_args.add_argument(
-        "--max-data-size", type=int, help="Maximum number of data points to load."
-    )
-    data_args.add_argument(
-        "-n",
-        "--num-workers",
-        type=int,
-        default=8,
-        help="Number of workers for the parallel data loading (0 means sequential).",
+        "-s", "--smiles-columns", nargs="+", help="The columns in the input CSV containing SMILES strings. If unspecified, uses the the 0th column."
```
[the remark this answers is not in the document]
Comment (JacksonBurns) [reply chemprop/cli/common.py]: Oops thanks only meant to push the change lines not all the reformatting. Should be fixed now

**Comment by davidegraff** [comment]
I'm a bit of a broken record at this point, but IMO this features falls under "reinventing the wheel." Users can already control GPU visibility via the `CUDA_VISIBLE_DEVICES` environment variable. Gaving them specify a specific GPU index via a CLI argument is only _marginally_ less work than temporarily setting the visibility:
```
$ CUDA_VISIBLE_DEVICES="0,2" chemprop train [...] --n-gpu -1
```
vs. (new)
```
$ chemprop train [...] --gpu-spec 0 2
```
or using _no_ GPUs:
```
$ CUDA_VISIBLE_DEVICES="" chemprop train [...]
$ # or `chemprop train [...] --n-gpu 0`
```
vs. (new)
```
$ chemprop train [...] --gpu-spec -2
```
I would really like to avoid having a CLI argument that is already covered by external behavior. Users will need to learn the syntax of one of the two options, but it would be better if we just forced them to use the general solution.

File: chemprop/cli/common.py [hunk 2]
```diff
@@ -19,13 +21,9 @@ def add_common_args(parser: ArgumentParser) -> ArgumentParser:
     data_args.add_argument(
         "-r",
         "--reaction-columns",
-        nargs="+",
+    parser.add_argument("-g", "--gpu-spec", nargs="+", type=int, default=-1, help="indexes of GPUs to use (-1 for all (default), -2 for CPU)")
         help="The columns in the input CSV containing reactions.",
     )
```
Comment (hwpang) [inline chemprop/cli/common.py:26]: I think some artifacts happen here that insert the changes into the `add_argument` block

**Comment by davidegraff** [comment]
Also, Lightning can auto-select underutilized GPUs if you only tell it the number that you want. If we stick to the paradigm of having users select _indices_, then we lose that capability. I would imagine that users broadly want "a given number of '_free_' GPUs" when they request resources, but it's unreasonable to expect that users actually know which GPUs are free.

File: chemprop/cli/common.py (hunk 2 above)
Comment (JacksonBurns) [reply chemprop/cli/common.py:26]: Bleh okay I will do this the right way and just change my formatter haha

**Comment by davidegraff** [comment]
I think what we might want to consider is abolishing the CLI concept of a GPU entirely and moving our parameters entirely over to the Lightning spec: `--accelerator` and `--devices`. We could then just point users to the Lightning documentation for more details.

**Comment by JacksonBurns** [comment]
> I'm a bit of a broken record at this point, but IMO this features falls under "reinventing the wheel." Users can already control GPU visibility via the `CUDA_VISIBLE_DEVICES` environment variable. Gaving them specify a specific GPU index via a CLI argument is only _marginally_ less work than temporarily setting the visibility:
> ...
> I would really like to avoid having a CLI argument that is already covered by external behavior. Users will need to learn the syntax of one of the two options, but it would be better if we just forced them to use the general solution ("teach a person to fish" and all that).

This is nice on UNIX but can't be done on Windows AFAIK. Windows does not allow per-line environment variables and requires setting them globally, calling the command, and then unsetting (see https://stackoverflow.com/a/1422082).

Agreed that teaching how to use these would be 'better', but I think to really hammer home the user-friendliness of chemprop we should expose everything through the CLI.

> Also, Lightning can auto-select underutilized GPUs if you only tell it the number that you want. If we stick to the paradigm of having users select _indices_, then we lose that capability. I would imagine that users generally want "a given number of '_free_' GPUs" when they request resources, but it's unreasonable to expect that users actually know which GPUs are free.

Where is this described in the lightning docs? I would like to read more.

**Comment by JacksonBurns** [comment]
> [This answer](https://stackoverflow.com/a/69758192) in your linked post describes a way to do exactly that and is thematically similar to UNIX in-line assignment.

Subshells don't inherit the conda environment, requiring this:
```pwsh
(chemprop) PS C:\Users\me> python -c "import torch"
(chemprop) PS C:\Users\me> pwsh -Command { python -c "import torch" }
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'torch'
(chemprop) PS C:\Users\me> pwsh -Command { conda activate chemprop; python -c "import torch" }
(chemprop) PS C:\Users\me>
```

> I'm also under the impression that Windows users can these days opt to just use WSL rather than native powershell, enabling them to avoid this entirely.

This seems like a _lot_ more effort for end users than it does for us to just expose this through the CLI.

> 
> At the point that we're releasing a command line application, is it not reasonable to ask a user be comfortable operating in a command line environment?

I don't think so. This tool is used by domain scientists, some number of whom will have never used a command line before. I don't think its unreasonable to cater to them.

> See [here](https://lightning.ai/docs/pytorch/stable/accelerators/gpu_basic.html#find-usable-cuda-devices)

Thanks!

**Comment by kevingreenman** [comment]
I don't think the "user-friendliness" question matters too much as long as we document the way to do it (IMO, `CUDA_VISIBLE_DEVICES="0,2" chemprop train [...] --n-gpu -1` and `chemprop train [...] --gpu-spec 0 2` are the same level of difficulty). 

If it's true that there's additional steps required for Windows users to use the environment variable solution, then I'm more inclined to support the new CLI option.

However, I think the auto-selection of underutilized GPUs is a point against the new CLI option and warrants more discussion.

**Comment by JacksonBurns** [comment]
Note for myself for later: using the visible devices variable may cause issues on clusters running slurm (which use it for their own purposes), see: https://slurm.schedmd.com/gres.html#GPU_Management

**Comment by JacksonBurns** [comment]
Another note: the source code for find usable devices doesn't take into account utilization of GPUs, just if it can access them or not (and returns the first n gpus it can access, n being the number the user requested) - see https://github.com/Lightning-AI/pytorch-lightning/blob/master/src%2Flightning%2Ffabric%2Faccelerators%2Fcuda.py#L83

File: chemprop/cli/common.py [hunk 3]
```diff
@@ -148,4 +153,6 @@ def process_common_args(args: Namespace) -> Namespace:
 
 
 def validate_common_args(args):
-    pass
+    # cannot use GPUs if torch cannot detect them
+    if args.gpu_spec != -2 and not torch.cuda.is_available():
```
Comment (davidegraff) [inline chemprop/cli/common.py:157]: In your current argparse model, it's not possible for `gpu_spec` to take a value of `-2`. It's either `-1` (default) or `list[int]`. What would happen in the case that a user supplies `-g -2 -1 0 1`?

File: chemprop/cli/common.py [hunk 4]
```diff
@@ -148,4 +153,6 @@ def process_common_args(args: Namespace) -> Namespace:
...
+        logger.error("Requested GPU(s) %s but torch does not detect GPU.", str(args.gpu_spec))
```
Comment (davidegraff) [inline chemprop/cli/common.py:158]:
```suggestion
        logger.error(f"Requested GPU(s) {args.gpu_spec} but torch does not detect GPU.")
```

File: chemprop/cli/predict.py [hunk 5]
```diff
@@ -247,8 +247,8 @@ def main(args):
         trainer = pl.Trainer(
             logger=False,
             enable_progress_bar=True,
-            accelerator="auto",
-            devices=args.n_gpu if torch.cuda.is_available() else 1,
+            accelerator="cpu" if args.gpu_spec == -2 else "auto",
```
Comment (davidegraff) [inline chemprop/cli/predict.py:250]:
```suggestion
            accelerator="cpu" if args.gpu_spec == -2 else "gpu",
```
small edge case, but there are many possible accelerators that could be selected sometimes (e.g., MPS can be selected on a GPU-enabled MAC)

File: chemprop/cli/train.py [hunk 6]
```diff
@@ -769,8 +768,8 @@ def main(args):
         trainer = pl.Trainer(
             logger=trainer_logger,
             enable_progress_bar=True,
-            accelerator="auto",
-            devices=args.n_gpu if torch.cuda.is_available() else 1,
+            accelerator="cpu" if args.gpu_spec == -2 else "auto",
```
Comment (davidegraff) [inline chemprop/cli/train.py:771]:
```suggestion
            accelerator="cpu" if args.gpu_spec == -2 else "cpu",
```

**Comment by KnathanM** [comment]
> I think what we might want to consider is abolishing the CLI concept of a GPU entirely and moving our parameters entirely over to the Lightning spec: `--accelerator` and `--devices`. We could then just point users to the Lightning documentation for more details.

Did you have thoughts about this @JacksonBurns? I feel like that is the most straightforward. Only downside is devices can be a list, int, or str, while argparse takes everything as a string, so multiple devices have to be passed as a single string like "1, 2". 
```
parser.add_argument("--accelerator", default="auto", help="Passed directly to the lightning Trainer().")
parser.add_argument("--devices", default="auto", help="Passed directly to the lightning Trainer(). If specifying multiple devices, must be a single string of comma separated devices, e.g. '1, 2'.")

Trainer(accelerator=args.accelerator, devices=args.devices)
```

**Comment by KnathanM** [comment]
After talking with Jackson we decided to punt this to v2.1 to have more time to get user input on the gpu flags. Just exposing `--accelerator` and `--devices` to the user is probably what we will do, but in any event the current v2/dev implementation uses `accelerator="auto"` and the user can choose the number of gpus via `-g` which should handle most immediate use cases.

**Comment by JacksonBurns** [comment]
TODO:
 - [ ] discuss cc58226 until we converge
 - [ ] if above (or a near approximation) is accepted, update the docs to (1) point users toward lightning's [`accelerator`](https://lightning.ai/docs/pytorch/stable/common/trainer.html#accelerator) and [`devices`](https://lightning.ai/docs/pytorch/stable/common/trainer.html#devices) documentation (or possibly [this](https://lightning.ai/docs/pytorch/stable/accelerators/gpu_basic.html#choosing-gpu-devices)) (2) update the accepted args
 - [ ] update the tests (?) which should fail now

**Review by KnathanM — APPROVED** [review]
In our last meeting, we talked about this PR. We think it is good to just expose accelerator and devices and point users to the using GPU lightning help page. 

I'll add a note here for later: This raises the question of how much of the lightning CLI we should implement. We decided that as a rule of thumb, we don't need to expose too much through our CLI. If the user wants to do something fancy with the lightning.pytorch.Trainer, they can use Chemprop from a notebook. The CLI is mostly for making it easy for new users to get started with Chemprop and to give some code examples for common workflows. Future additions to our CLI can be considered in a case by case basis.

**Comment by KnathanM** [comment]
In terms of documentation, maybe it belongs under the Command Line Tutorial [(this page)](https://chemprop.readthedocs.io/en/v2-dev/tutorial/cli/index.html) and could be something like:

Chemprop is built on top of Lightning, which has support for training and predicting on GPUs. Relevant CLI flags include `--accelerator` and `--devices`. See the [Lightning documentation](https://lightning.ai/docs/pytorch/stable/accelerators/gpu_basic.html#choosing-gpu-devices) and CLI reference for more details.

**Pull request merged**